RFC 15: Revisions Admin API (Revised)

marcofucci opened the pull request:
This RFC proposes an extension to the Admin API to allow a page to be published and its revisions to be retrieved, created and managed.

kaedroho commented:
Thanks for writing this up!

Given there is already an accepted RFC (#11) that covers this exact feature, I think we should consider this as an amendment to that RFC rather than just accepting something new ignoring it ever existed. Therefore, to kick off the discussion, I'll list out everything I see is different and explain my design decisions when I wrote RFC 11.

**Changing the URL format for page revisions**

The URL for getting revisions of a page has be changed from:

``GET /api/pages/revisions/?page_id=1``

to:

``GET /api/pages/1/revisions/``

This is the format I went with when I initially proposed RFC 11 [but I changed it to a GET](https://github.com/wagtail/rfcs/pull/11/commits/5a22e4b5d1b2a96b8cdd26cc010fc7960ee4e5b9) parameter as this would allow querying revisions across different pages (for uses like getting all revisions submitted by a particular user, or get all revisions awaiting moderation).

**There are no author or date filters**

It's unclear whether this RFC is suggesting we should ditch these features or not. I added these for two features we may add in the future:

The date filter stuff was created to make it easy to find the state of the page on a particular date which is something I think would be very useful for a future enhancement to the revisions diff API. I made it use date ranges since this as this would be the most flexible option for future enhancements after that.

The author filter was added for the "find all recent edits" by this author.

**Method of submitting for moderation**

The new RFC adds a boolean field that you can set to true to "submit" the revision and false to subsequently "reject" it. RFC 11 has an optional ``submit_for_moderation`` parameter on the "save a new revision" endpoint to submit to moderation but doesn't provide a way to "reject" it.

**Publishing a page**

The new RFC publishes a page by performing a ``POST`` request to ``/api/pages/<page-id>/publish/?revision-id=<id>``. RFC 11 has an optional ``publish`` parameter on the "save a new revision" endpoint which publishes the page if true.

**Preventing double-edits from occuring**

RFC 11 contains a method to prevent edit conflicts, the new RFC explictly excludes this to keep the RFC simple.

kaedroho commented:
Here's my own opinions on these changes:

> Changing the URL format for page revisions

I'm on the fence on this one. I do prefer the look of the RFC15 URLs but the RFC11 URLs allow the endpoint to be used to drive features outside of the page editor. I'll leave this one for everyone else to decide.

> Method of submitting for moderation

Submitting for moderation has side-effects (it sends emails) so I'm really not a fan of allowing this to be controlled by changing the value of a field. I think either adding ``?submit_for_moderation=true`` to the save view or adding a new endpoint (eg ``/api/pages/revisions/<id>/submit_for_moderation/``) would both be neater.

> Publishing a page

The method suggested in this new RFC appears to allow any revision to be published, even if it's an old one. In Wagtail, we always create a new revision on publish (so reverting to an old revision will copy that content into a new revision and publish it). I think allowing any revision to be published without copy could be confusing for administrators as the revisions UI will not pick this up.

> Preventing double-edits from occuring
> ...the new RFC explictly excludes this to keep the RFC simple.

Fair enough :)

marcofucci commented:
@kaedroho thanks for this and sorry for the duplication, the `RFC 11` has not been moved to the accepted folder yet and I missed the comments on the related merged PR.

As that PR has already been merged, I'm wondering what's the best way to proceed.

We could:
1. close this and open a new PR with changes on the original `RFC 11` instead which means that the `RFC 11` will be re-discussed and implemented.
2. keep this `RFC 15` as an updated version of the `RFC 11` which means that the `RFC 15` will be discussed and implemented.

I don't have any preferences so I'll leave it to you.

Thanks!

kaedroho commented:
> the RFC11 has not been moved to the accepted folder yet

Ahh, yes. Sorry about that, it should've been moved there when it was merged.

marcofucci commented:
@kaedroho if you let me know which option you prefer, I'll tidy things up and continue the discussion.

marcofucci commented:
I updated this PR with the following changes:

- I've added the _Data format_, _URL structure_ and _Abstract_ sections from the original RFC 11
- I've reverted back to specifying pages as GET parameters
- I've added a section on filtering by author as in the original RFC 11
- I haven't included the section on filtering by dates as it's not needed yet
- I've added a section on creating the first revision of a new page
- I haven't included the section on getting a revision by earliest/latest created_at date. I suggest we could use`/api/pages/revisions/head/?page_id=<page-id>` instead.
- I've added a double-edit section specifying that it will not be part of this RFC
- I've added back the section on _Locked pages_ from the original RFC 11

Regarding the other sections, it's usually better not to use parameters as actions so we could try with custom POST actions instead.

How about:

**Submit for moderation:**
`POST /api/pages/revisions/<revision-id>/moderation/submit/`

**Reject moderation:**
`POST /api/pages/revisions/<revision-id>/moderation/reject/`

**Publishing a page:**
We could either use a `from_revision_id` param:
`POST /api/pages/publish/?page_id=<page-id>&from_revision_id=<revision-id>`
and return the newly created revision id. 

Or use:

`POST /api/pages/revisions/?page_id=<page-id>` 
passing `live=True` in the body.

**Unpublish a page**
We would still need this endpoint so I suggest we keep
`POST /api/pages/unpublish/?page_id=<page-id>`

@kaedroho please let me know what you think

kaedroho commented:
Thanks @marcofucci, this is looking good!

The ``?page_id`` parameter is only relevant for the revisions listing view, where it's only used to filter the list of revisions. I don't think it should be used when you are working on a page instance itself.

So, I think ``POST /api/pages/<page-id>/unpublish/`` is nicer than ``POST /api/pages/unpublish/?page_id=<page-id>`` and ``POST /api/pages/<page-id>/publish/?from_revision_id=<revision-id>`` is nicer than ``POST /api/pages/<page-id>/publish/?from_revision_id=<revision-id>``

> I haven't included the section on getting a revision by earliest/latest created_at date. I suggest we could use/api/pages/revisions/head/?page_id=<page-id> instead.

Would this return the revision or redirect to it?

I think it's quite nice to have a view which takes you directly to the latest revision, but I really don't think the URLs look nice here (the page id should really be before "head" somewhere).

This is a good argument for the URL structure that you initially proposed, but I think there's a middle-ground between the two we could reach. How does the following look to you:

Get all revisions: ``GET  /api/pages/revisions/``
Get revisions for a specific page: ``GET  /api/pages/<page-id>/revisions/``
Create a new revision: ``POST  /api/pages/<page-id>/revisions/``
Get latest revision for a specific page: ``GET  /api/pages/<page-id>/revisions/head/``
Get an individual revision: ``GET  /api/pages/revisions/<revision-id>/``
Submit a revision for moderation: ``POST  /api/pages/revisions/<revision-id>/submit/``

However, the following will not work:

``GET  /api/pages/<page-id>/revisions/<revision-id>/``
``POST  /api/pages/<page-id>/revisions/<revision-id>/submit/``

> How about:
> Submit for moderation:
> POST /api/pages/revisions/<revision-id>/moderation/submit/

I prefer this over giving access to the underlying ``submitted_for_moderation`` boolean.

I think one thing we should be mindful of is Wagtail's editor always creates a new revision when publishing or submitting for moderation and this proposed API would allow submitting/publishing an old revision without creating a new one. I'm not sure if this incorrect behaviour (ping @gasman) but if it is, we shouldn't allow it in the API.

marcofucci commented:
@kaedroho we have changed almost all the endpoints after your comment and the following considerations:

`page-id` as part of the URL is better than as GET param.

Always having `page-id` as part of the URL is better than having it only in some cases.
We propose using a special character `'-'` instead of `page-id` when the id is not needed as described in the [Google API Design Patters](https://cloud.google.com/apis/design/design_patterns#list_sub-collections).
This makes the URL format clearer, simpler and it also makes it easier to write DRF routers.

The moderation/publishing endpoints are now disconnected from specific revisions and they create new revisions.
This is also how Wagtail works internally so it will be easier to implement.

kaedroho commented on draft/015-revisions-api-revised.rst:
```diff
@@ -71,6 +71,38 @@ The endpoints are grouped as some relate to revisions and some to pages.
+
+.. code-block:: http
+
+    GET /api/pages/revisions/
```
Why does this URL not include the ``-``?

marcofucci replied:
I completely agree. I wanted to check what you thought about the `-` before changing this as well.

kaedroho commented:
I'm happy with adding the ``-``. Looks like a couple of the URLs haven't got this and not sure if that's a typo.

kaedroho commented on draft/015-revisions-api-revised.rst:
```diff
@@ -0,0 +1,264 @@
+
+.. code-block:: http
+
+    GET /api/pages/<page-id>/revisions/<revision-id>/
```
What is the canonical defail URL of a revision?

is it ``/api/pages/-/revisions/<revision-id>/``

or is it ``/api/pages/<page-id>/revisions/<revision-id>/``

I'd prefer that we choose one and block the other.

marcofucci replied:
I would say `/api/pages/<page-id>/revisions/<revision-id>/` as it's more explicit, what do you think?

kaedroho replied:
I agree

kaedroho commented on draft/015-revisions-api-revised.rst:
```diff
@@ -0,0 +1,264 @@
+
+    GET /api/pages/<page-id>/revisions/<revision-id>/
+
+This returns an error if `<revision-id>` does not refer to `<page-id>`.
```
404 error?

I think this could be a little clearer:

> This returns a 404 error if `<revision-id>` does not reference a revision that belongs to the page

marcofucci replied:
Agreed

kaedroho commented on draft/015-revisions-api-revised.rst:
```diff
@@ -0,0 +1,264 @@
+    GET /api/pages/<page-id>/revisions/head/
+
+This redirects to ``/api/pages/<page-id>/revisions/<revision-id>/``
+
```
What happens if I try ``GET /api/pages/-/revisions/head/``?

marcofucci replied:
I would say, you get the latest revision without filtering by page still redirecting to `/api/pages/<page-id>/revisions/<revision-id>/` as the other.
It wouldn't be very useful and not likely used but at least we don't break this specs.
What do you think?

kaedroho replied:
This sounds sensible to me

kaedroho commented on draft/015-revisions-api-revised.rst:
```diff
@@ -0,0 +1,264 @@
+
+This will reject the revision awaiting moderation.
+
+
```
How are revisions in moderation accepted? This can't be done with ``/publish`` because the revision in moderation doesn't have to be the latest revision

marcofucci replied:
`/publish` expects the content as body and creates a new revision the same way as _Create a new revision_ one but making it live as well.
Therefore, I thought we could accept a revision in moderation by publishing a new one using the `/publish` endpoint.

It's up to the client to send the content of the revision awaiting moderation as body to the publish endpoint.

kaedroho replied:
I see, thanks. I guess I got confused as the "save revision" URL is ``POST /api/pages/<page-id>/revisions/``. I think we should think about how we make these publishing URLs clearer.

I know it's already been discussed, but I think adding the action as a GET parameter would be the clearest option:

Save revision: ``POST /api/pages/<page-id>/revisions/``
Save revision and publish it: ``POST /api/pages/<page-id>/revisions/?action=publish``
Save revision and submit it for moderation: ``POST /api/pages/<page-id>/revisions/?action=submit-for-moderation``

This is basically how Wagtail's edit view currently works

marcofucci replied:
Just to understand, how would you reject a revision and unpublish a page?
Would you still use `POST /api/pages/<page-id>/moderation/reject/` and `POST /api/pages/<page-id>/unpublish/` ?

kaedroho replied:
Yep, ``POST /api/pages/<page-id>/unpublish/`` should definitely be the URL to unpublish a page as this action doesn't do anything with revisions.

Not sure what the best thing to do with ``reject`` is. Within Wagtail, accept and reject are both actions that are performed on a specific revision (they are methods on the ``Revision`` model and also the accept/reject views within wagtail take a revision ID)

If we were to follow that, I think this would mean doing the accept/reject actions through two URLs within the revision:

`POST /api/pages/<page-id>/revisions/<revision-id>/moderation/accept/  (takes no body content, publishes this revision's content)`
`POST /api/pages/<page-id>/revisions/<revision-id>/moderation/reject/`

zerolab replied:
Throwing noise in the mix -- should it not be PUT or PATCH for accept/reject?

marcofucci replied:
@zerolab what field would the accept/reject endpoint  PATCH?

zerolab replied:
as far as the convention goes:
- `POST` creates a new resource
- `PUT` updates a resource
- `PATCH` is a delta update.

In this cars accepting/rejecting a revision changes their state to published, so one could argue that it is the 'status' field (on an abstract level).

Do ignore the curveball. This is mostly nitpicking..

marcofucci replied:
@zerolab this is similar to what we started with, check the [first version of this RFC](https://github.com/nhsuk/wagtail-rfcs/blob/ee357caa0facde0e0eb792b2666484eddbd65c21/draft/012-manage-page-changes-api.rst#submit-for--reject-moderation).

I also prefer PATCHing a field (`status` or `submitted_for_moderation`) but I had the impression that this would have some internal consequences so I'm happy with POST actions as well.

In general `POST` methods are used to create new resources and for custom actions as well so that shouldn't be too much of a problem. I believe it's advisable not to use too many custom actions though.

Can we first start with checking if we all agree with these:
```
Save revision: `POST /api/pages/<page-id>/revisions/`

Save revision and publish it: `POST /api/pages/<page-id>/revisions/?action=publish`

Save revision and submit it for moderation: `POST /api/pages/<page-id>/revisions/?action=submit-for-moderation`

Unpublish a page: `POST /api/pages/<page-id>/unpublish/`
```

and we are only discussing the accept/reject endpoints?

marcofucci commented:
@kaedroho @zerolab 

**Save / Submit for moderation / Publish**
I've updated the save endpoints as discussed.
I've used `post_save` instead of the word `action` as the action is really _save_ but happy
to change this to a better word or use `action` if you prefer.

**Accepting / Rejecting**
_I believe this is the last bit to be discussed before this RFC is ready._

Regarding accepting and rejecting a specific revision, I understand this is how Wagtail works.
I've noticed that when a specific revision is awaiting moderation, you can still change the page
and create further revisions which, if saved as draft, diverge from the version in moderation.

We haven't done extensive user research on this yet but it seems quite a confusing concept
to understand for an editor as you are not sure at a particular point in time what version you
are moderating, what is in draft, what is live and which version you are changing.

I understand that this is to avoid blocking the editing process and to allow editors to keep
changing the page whilst a previous version is awaiting moderation.

However, I don't quite get why you wouldn't want to always moderate the latest revision.

In my mind, let's say that Alice submits the revision 2 for moderation and asks Bob to review it.
If Alice forgets a paragraph and saves the page (creating a new revision 3), she surely
wants revision 3 to be reviewed by Bob without the annoyance of remembering to 
submit the new revision for moderation.

If this is true, I think the status of `submitted for moderation` should really be
on a page and not on a revision from a high level point of view.
This doesn't exclude keeping the flag on the revision object and update it via code to make it
look like it's on the page object instead.

It might be a different discussion but this has direct impact on how we design the API
as the endpoints `/api/pages/<page-id>/revisions/<revision-id>/moderation/accept/` that
we considered a few times reference specific revisions.

What do you think?

kaedroho commented on draft/015-revisions-api-revised.rst:
```diff
@@ -0,0 +1,240 @@
+
+.. code-block:: http
+
+    POST /api/pages/-/revisions/
```
Creating a new page requires us to also know the page type and the ID of the page we're creating under. Neither of these are in the revision though so I think we should change this to something like:

``POST /api/pages/<parent-page-id>/add_child/?type=blog.BlogPage``

But like the other save-revision views, you would only need to submit the value of the content here. For example:

```
POST /api/pages/<parent-page-id>/add_child/?type=blog.BlogPage 
Content-Type: text/json

{
    "title": "My new blog post",
    "body": "Blog post here"
}
```

gasman commented:
> However, I don't quite get why you wouldn't want to always moderate the latest revision.

In short, race conditions: it ensures that changes don't sneak in between the moderator looking at a page and hitting 'approve'.

More generally, though, I think it would be risky to break the principle that things only go live on the site after you (or someone else) has explicitly decided to do so by hitting publish or submit.

kaedroho commented:
Thanks again @marcofucci 

> I've used ``post_save`` instead of the word ``action``

I would *slightly* prefer to use ``after_save`` over ``post_save``. The word "post" makes me think of the HTTP verb!

How about the word ``then``? I think this reads quite nicely: ``?then=publish`` ``?then=submit-for-moderation``.

marcofucci commented:
Thanks @gasman, I understand what you mean and yes I see the point of designing API that accept / reject specific revisions now. It's flexible enough to allow us to implement our own moderation flow if necessary as there are cases were you might want a github PR like approach.

@kaedroho I like `then`.

marcofucci commented:
Before discussing the creation of a page, I would like to agree on the approving / rejecting bit first.

I would prefer using `PATCH` for these cases but I understand that they do more than just changing states and we might want to extend them in the future (e.g. adding a comment when rejecting) so I think using custom actions might be OK.

Would these be OK?
`POST /api/pages/revisions/<revision-id>/moderation/approve/`
`POST /api/pages/revisions/<revision-id>/moderation/reject/`

kaedroho commented:
@marcofucci Yep, those look OK to me!

marcofucci commented:
The last thing to discuss I think it creating a new page.

It seems like the main problem is that `parent` and `type` are properties of the page, not revision.
I think it's probably a good idea not only to allow users to create but also change these fields (e.g. to move pages).

I can't think of anything which I particularly like and the `add_child` only works when creating
pages so open to suggestions.

How about:

**1**: `POST /api/pages/` creates a page and a revision. 
You need to send `meta` and `revision` in the body:
```
{
	'meta': {
		'parent_id': <parent-id>,
		'type': <page-type>
	},
	'revision': {
		'title': 'lorem ...'
	}
}
```

**2**: `PATCH /api/pages/<page-id>/meta/` to patch the `parent` field, maybe the `type` as well and other
fields in the future.

**3**: We could even have a `GET /api/pages/<page-id>/meta/` which only returns the meta data.

This plays well with `/api/pages/<page-id>/revisions/` with all the other endpoints we have defined.

kaedroho commented:
> POST /api/pages/ creates a page and a revision.

Happy with this

> You need to send meta and revision in the body:

Putting the content in a "revision" key is different to the way that pages are represented elsewhere in the API. I would prefer to keep the representation the same, unless there is good reason not to.

> PATCH /api/pages/<page-id>/meta/ to patch the parent field, maybe the type as well and other
fields in the future.

This got me thinking. So far, I've been thinking of the ``meta`` section as "anything that doesn't get displayed on the page in the frontend" (with the exeption of the ``id``). But this definition doesn't make much sense for the admin API.

So I think it might make sense to redefine ``meta``, just for the admin API, as anything that shouldn't be captured in a revision. So this would be just the following fields:

 - The pages id
 - The content type
 - The tree position
 - The ``url_path``, ``live``, has_unpublished_changes``, `owner``, ``locked``, ``latest_revision_created_at`` and ``first_published_at`` fields
 - Anything that's not in the database, eg ``detail_url`` and ``html_url``

The value of the revisions ``content`` field would then be the same as its page object minus the ``meta`` section, which would probably simplify some client code.

This is probably a little out of scope for this RFC though :)

marcofucci commented:
I think we can try and have _meta_ as close as possible to the meta you currently get back when you GET a page (I'm looking at http://admin-api-demo.herokuapp.com/admin/api/v2beta/pages/10/), we might need to add a few more props but `slug`, `type` and `parent` are there.

> Putting the content in a "revision" key is different to the way that pages are represented elsewhere in the API. I would prefer to keep the representation the same, unless there is good reason not to.

I agree that putting the content in a revision key is a bit different. I guess I was trying to namespace fields based on the API level.

Level 1 endpoint `/api/pages/` with _meta_ and _revision_.
Level 2 endpoint `/api/pages/<page-id>/meta/`
Level 2 endpoint `/api/pages/<page-id>/revisions/`

All other endpoints are under `.../revisions` so they don't need the `revision` key.

Another option is to just lose the `revision` key and have:

```
{
	'meta': {
		'parent_id': <parent-id>,
		'type': <page-type>
	},
	'title': 'lorem ...'
}
```

so that it matches the response you get back from the GET page API.

kaedroho commented:
I'd still prefer to set the meta fields over the URL. This would make the request body contain only content and would reflect exactly what goes in to the new revision too.

Having a clear difference between the metadata and content like this would make certian actions, such as copying and restoring a revision, much easier to implement on the client side as you wouldn't need to know which fields are metadata and need to be stripped out (otherwise you might accidentally move the page, for example).

I think this also gives us more flexibility for how we can position the new page:

``POST /api/pages/1/add_child/?type=blog.BlogPage&position=first`` (defaults to ``last``)

We could even add sibling pages:

``POST /api/pages/1/add_sibling/?type=blog.BlogPage&position=previous`` (defaults to ``next``)

marcofucci commented:
I understand your point and I had a quick look at the Revision / Page models.

I see two issues:

**What is metadata and what is not**
The db record of a revision includes all the page fields, even the ones not related to a particular revision (e.g. `path`). This is already confusing for a developer as it's not clear what happens when you restore an old revision for example.
I might have missed something but it could be worth working on splitting the fields between meta and revision fields or always use revision data and move pages when restoring old revisions. Definitely not part of this PR though.

Also, the word _metadata_ or _meta_ might be confusing as we are already returning meta in the listing and I'm not sure if we want to align these two concepts or not.

We should try:
- decide what is metadata and what is not
- decide if this _meta_ should be the same as the listing _meta_ and if not, choose a different name

**A POST action per metadata field**
The `/api/pages/1/add_child/` endpoint allows you to specify the position only when creating a child. What happens when you need to move a page, would you have a different endpoint `/api/pages/1/move/`?
If each metadata field translates into 2 endpoints, we might end up with too many custom actions.

kaedroho commented:
I feel like the scope of this discussion is rapidly increasing, we should probably consider splitting some of this off into more RFCs so we can start work on some of the things that have been agreed so far.

I propose we keep this RFC focused on viewing and creating new revisions and move the discussion for creating of new pages, meta fields and new page actions into one or two new RFCs. What do you think?

Here's my comments on the two issues you've highlighted:

> What is metadata and what is not

I think we should redefine this for the whole admin API (but not the public API) as the current definition of "does it display on the frontend?" isn't that relevant for the admin.

I think defining this is by whether the field is versioned or not as this seems to be a good way to separate page content fields from Wagtail's internal "meta" fields.

That list of fields is currently hard-coded in the ``as_page_object`` method of ``PageRevision``: https://github.com/wagtail/wagtail/blob/da067679cdd7999c256c8916f78c82f7641d51cd/wagtail/wagtailcore/models.py#L1421-L1443

(@gasman should ``slug`` and ``show_in_menus`` be put in the unversioned fields as well?)

We should also include any generated fields that wholely rely on just meta fields. For example ``html_url``, ``detail_url``, ``children``, ``descendants``, etc.

> If each metadata field translates into 2 endpoints, we might end up with too many custom actions.

Different meta fields have different requirements. Here's what I think we could to to handle each field:

- The pages id: cannot change
- The content type: cannot change
- The tree position: ``POST /api/pages/<page-id>/move/``
- ``url_path``, ``latest_revision_created_at``, `live``, ``has_unpublished_changes``: Automatically updated by other actions and cannot be changed directly by users
- ``owner``, ``has_unpublished_pages``, ``locked``: Perhaps we can allow ``PUT``/``PATCH`` requests to ``/api/pages/<page-id>/meta/`` to update just these fields?
  - Maybe ``locked`` should have its own endpoint. For example ``PUT /api/pages/<page-ids>/edit-lock/`` and ``DELETE /api/pages/<page-ids>/edit-lock/``

So that's 2 or 3 actions to handle all of the metadata field changes.

gasman commented:
> (@gasman should `slug` and `show_in_menus` be put in the unversioned fields as well?)

No, I don't think so - why would they be? They're a meaningful part of the page state as it existed at that moment in time (unlike, say, `path` or `has_unpublished_changes`).

marcofucci commented:
@kaedroho I agree, I've removed the sections that mentioned the creation of new pages which will not be included in this RFC as they require more thinking.

@kaedroho @gasman @zerolab shall we go through this RFC one more time and see if there's anything else to discuss?

kaedroho commented on draft/015-revisions-api-revised.rst:
````````````````````diff
@@ -0,0 +1,236 @@
+revision that belongs to the page.
+
+By revision id only
+```````````````````
````````````````````
Would this view return the revision or 301 redirect to it's canonical URL (``/api/pages/<page-id>/revisions/<revision-id>/``)?

marcofucci replied:
Yes, I was questioning this as well and I didn't include anything because I'm OK with both.
I slightly prefer 301 with redirect to its canonical URL but this would requires 2 requests which is not ideal.

kaedroho replied:
I slightly prefer 301 as well, so maybe we should just go with that. Everywhere where we link to revisions, we would link to the canonical URL, so I don't think this URL would be used much anyway.

kaedroho commented:
This looks good to me 👍 

Just one little thing I think we should clarify: https://github.com/wagtail/rfcs/pull/15#discussion_r104621120

kaedroho commented:
I've submitted an RFC for all the meta changes discussed here: https://github.com/wagtail/rfcs/pull/18

kaedroho commented:
Could we remove the "Unpublish a page" section for now? It isn't actually related to revisions, I'll copy it over to a new RFC that deals with page actions in general.

kaedroho commented:
I've added a "page actions" RFC at https://github.com/wagtail/rfcs/pull/19

(I didn't get long to work on it this morning so it doesn't yet cover page creation)

kaedroho commented on draft/015-revisions-api-revised.rst:
```diff
@@ -0,0 +1,219 @@
+Create a new revision
+^^^^^^^^^^^^^^^^^^^^^
+
+Creating a new revision is done by submitting the value of the "content" field
```
As it stands, to create a new revision. I have to submit the value of the "content" field:

for example:
```
POST /api/pages/<page-id>/revisions/

{
    "title": "Foo"
}
```

But when I GET the created revision, the structure changes:

```
GET /api/pages/<page-id>/revisions/1/

{
    "id": 1,
    "created_at": "2017-03-10T10:21:00Z",
    "created_by": "karl",
    "content": {
        "title": "Foo"
    }
```

Maybe I'm being picky, but we should probably change this to:

```
POST /api/pages/<page-id>/revisions/

{
    "content": {
        "title": "Foo"
    }
}
```

Or allow a ``PUT``/``PATCH`` requests to ```/api/pages/<page-id>/content/`` (this would allow changing content by ``PATCH`` as well, which might be useful)

marcofucci replied:
@kaedroho first of all, what's the reason behind returning
```
{
    "id": 1,
    "created_at": "2017-03-10T10:21:00Z",
    "created_by": "karl",
    "content": {
        "title": "Foo"
    }
}
```

and not just
```
{
    "id": 1,
    "created_at": "2017-03-10T10:21:00Z",
    "created_by": "karl",
    "title": "Foo"
}
```

The main problem with the `content` field is that extra fields can be added by developers and they might not be all _content_ fields.

kaedroho replied:
I think it's good to have a distinction between the fields used to describe the revision from the fields that describe the revision's content. Much like how we have the meta/non-meta separation for pages currently.

This should improve clarity and make it easy for client libraries to automatically work out what field is for what (so if we add a new field to all revisions, it'll be nice if clients could tell that field was not page content).

kaedroho replied:
Perhaps this would be better?

```
{
    "meta": {
        "id": 1,
        "created_at": "2017-03-10T10:21:00Z",
        "created_by": "karl"
    },
    "title": "Foo"
}
```

kaedroho replied:
If https://github.com/wagtail/rfcs/pull/18 gets merged, this would make everything outside "meta" identical to pages. I think that would be quite neat as you can swap the "meta" section with the page's "meta" and you've built a valid representation of the page at that time.

kaedroho commented:
As @gasman mentioned in https://github.com/wagtail/rfcs/pull/19#issuecomment-285633747 we should mention authentication/permissions for the writable bits.

kaedroho commented:
I've removed the FCP label from this as there's a few unresolved questions we need to think about before we can implement it:

 - What the structure of the revision's JSON representation should be https://github.com/wagtail/rfcs/pull/15#discussion_r105369491
 - Should we serialise the page representation of a revision or just output the raw internal JSON? https://github.com/wagtail/rfcs/pull/15#discussion_r108897823
 - Authentication and permissions https://github.com/wagtail/rfcs/pull/15#issuecomment-285640999

kaedroho commented:
@marcofucci I've made a few tweaks in https://github.com/wagtail/rfcs/pull/15/commits/6f2260d18549c69a27ede52b30a70624f68031f4. I think this resolves the last remaining questions. Could you let me know if you're happy with those changes?

marcofucci commented:
👍  it looks good, @kaedroho !
